[master]Add Ballerina Deployment Guide (#9621)
ballerina-platform/ballerina-dev-website: swan-lake/development-tutorials/deployment-guide/k8s-deployment.md

title: Kubernetes deployment
description: When deploying Ballerina services in a Kubernetes environment, it is essential to focus on high availability and scalability. Key considerations include determining the appropriate number of worker nodes and replicas. These decisions help scale the deployment and ensure that the system remains available and responsive under load. 
keywords: ballerina, programming language, services, cloud, kubernetes, docker
active: ballerina-deployment-guideLines
intro: When deploying Ballerina services in a Kubernetes environment, it is essential to focus on high availability and scalability. Key considerations include determining the appropriate number of worker nodes and replicas. These decisions help scale the deployment and ensure that the system remains available and responsive under load. 
---

## Simple application deployment in Kubernetes

Ballerina simplifies Kubernetes artifact generation by deriving configurations directly from your code, enabling developers to focus on business logic. Below is a guide for deploying Ballerina applications on Kubernetes using Ballerina’s Code-to-Cloud capabilities.

### Ballerina application code

> Here is an example of a Ballerina service that serves a simple "Hello from Kubernetes!" message on an HTTP request:

Reviewer: ```suggestion
Here is an example of a Ballerina service that serves a simple `Hello from Kubernetes!` message on an HTTP request:
```

```ballerina
import ballerina/http;

// HTTP Listener on port 9090
listener http:Listener helloEP = new (9090);

> // Define the HTTP service

Reviewer: We don't need this comment IMO

Expose the service via NodePort for local testing

```
$ kubectl expose deployment hello-deployment --type=NodePort --name=hello-svc-local
service/hello-svc-local exposed
```

Get the Kubernetes service details
```
$ kubectl get svc

NAME                        TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)          AGE
hello-svc               ClusterIP       10.97.140.84        <none>        9090/TCP         13m
hello-svc-local         NodePort        10.108.87.21        <none>        9090:31360/TCP   2m18s
```
Access the service using the external IP and NodePort

For Minikube users, you can obtain the IP using:
```yaml
$ minikube ip
```
> **Tip:** you can access your service locally via `kubectl port-forward`
 `$ kubectl port-forward svc/hello-k8s-svc  8081:8080`

Test the service using curl

```
$ curl http://<EXTERNAL-IP>:<NODE-PORT>/helloWorld/greeting
```

## Kubernetes deployment with config maps and autoscaling

For a more complex deployment, this demonstrates a Ballerina app that reads a greeting from a config map and responds to HTTP requests. It covers Kubernetes deployment with resource limits, autoscaling, config maps, and health probes using Code to Cloud. 

> 1. Create a new Ballerina package

Reviewer: ```suggestion
1. Create a new Ballerina package.
```

Reviewer: Check all the places.

```bash
$ bal new hello_k8s
```
2. Replace the `main.bal` file with the following content
```ballerina 
import ballerina/http;

listener http:Listener helloEP = new(9090);

configurable string greeting = ?;

service /helloWorld on helloEP {
    resource function get greeting() returns string {
        return greeting + ", Kubernetes!";
    }

> }

Reviewer 2: @lnash94 please check the EOL format in the `md` file. It seems to have changed to `CRLF`. Change it to LF and commit.

> }

Author: Removed the trailing white spaces in the language code.

```
3. Create `probes.bal` with health and readiness probes

```ballerina
import ballerina/http;

listener http:Listener probeEP = new(9091);

service /probes on probeEP {
    resource function get healthz() returns boolean {
        return true;
    }
    resource function get readyz() returns boolean {
        return true;
    }
}
```
4. Create a `Config.toml` file

```toml
greeting = "Hello"
```

5. Update the `Cloud.toml` to include auto-scaling, resource limits, config maps, and probe configuration

```toml
[container.image]
repository="wso2inc"
name="hello-k8s"
tag="v0.1.0"

[cloud.deployment]
min_memory="100Mi"
max_memory="256Mi"
min_cpu="500m"
max_cpu="500m"

[cloud.deployment.autoscaling]
min_replicas=2
max_replicas=5
cpu=60

[[cloud.config.files]]
file="./Config.toml"

[cloud.deployment.probes.liveness]
port=9091
path="/probes/healthz"

[cloud.deployment.probes.readiness]
port=9091
path="/probes/readyz"
```

> 6. Build the package: 

Reviewer: ```suggestion
6. Build the package.
```

```
$ bal build --cloud="k8s"
  Compiling source
       wso2/hello_k8s:0.1.0

  Generating executable

  Generating artifacts...

       @kubernetes:Service                      - complete 1/2
       @kubernetes:Service                      - complete 2/2
       @kubernetes:ConfigMap                    - complete 1/1
       @kubernetes:Deployment                   - complete 1/1
       @kubernetes:HPA                          - complete 1/1
       @kubernetes:Docker                       - complete 2/2 

       Execute the below command to deploy the Kubernetes artifacts: 
       kubectl apply -f /home/wso2/user/hello_k8s/target/kubernetes/hello_k8s

       Execute the below command to access service via NodePort: 
       kubectl expose deployment hello-k8s-deployment --type=NodePort --name=hello-k8s-svc-local

       target/bin/hello_k8s.jar
   ```

> > **Tip:** You can update `Ballerina.toml` with build option to integrate option to `bal build`

Reviewer: suggestion:

You can modify the `Ballerina.toml` file to include a build option, allowing the integration of the bal build command with specific configurations.

> ```
> [build-options] 
>  cloud = "k8s"
> ```
   

> - Execute the Kubernetes service, Push Docker image to Docker Hub: 

Reviewer: This bullet point is not makes sense it seems as it is not having any relationship with next bullet point. `Build the GraalVM executable in Kubernetes`

Author: I rearranged this `Push Docker images` point as step 7. 

> step 6. Build the package.
> step 7.  Push Docker images.
> Step 8.  Deploy the Kubernetes artifacts.

and Shall I arrange this graalVm executable as a tip for the `Build the Package` sections?

7. Deploy the Kubernetes artifacts 
```
 $ kubectl apply -f target/kubernetes/hello_k8s
```
8. Testing the Kubernetes Deployment
   Verifying the Kubernetes deployment ensures that the application is running correctly and meets the expected functionality within the Kubernetes environment.
  - Verify the pods
    ```
    $ kubectl get pods
    ```
  - Expose the service via NodePort
    ```
    $ kubectl expose deployment hello-k8s-deployment --type=NodePort --name=hello-k8s-svc-local exposed
    ```
  - Get the External IP and port
    ```
    $ kubectl get svc
    NAME                      TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)          AGE
    hello-k8s-svc         ClusterIP   10.96.173.207   <none>        9090/TCP,9091/TCP               5m11s
    hello-k8s-svc-local   NodePort    10.99.245.41    <none>        9090:30342/TCP,9091:30515/TCP   66s
    kubernetes            ClusterIP   10.96.0.1       <none>        443/TCP                         130m
    ```
  - Execute the curl http://192.168.49.2:30342/helloWorld/greeting command to access the deployed service via cURL.
    ```
     $ curl http://192.168.49.2:30342/helloWorld/greeting
     Hello, Kubernetes!
    ```

>     Follow the below references for more information:

Reviewer: ```suggestion
    Follow the references below for more information:
```

>   - [Deploy Ballerina on Kubernetes](https://ballerina.io/learn/deploy-ballerina-on-kubernetes/)

Reviewer: Thse bullet points are not indented.

Author: fixed it

  - [Kubernetes - Hello world](https://ballerina.io/learn/by-example/kubernetes-hello-world/)

## Ballerina deployment with Kubernetes Kustomize

> [Kustomize](https://kustomize.io/) is a tool that allows you to modify Kubernetes YAML files without altering the original files. It can enhance the generated YAML from code-to-cloud deployments by applying additional customizations. The `kustomization.yaml` file in the root directory serves as a sample, combining the generated YAML files from different projects. Environment variables, such as specifying the location of the `Config.toml` for a service, can be added using Kustomize patches, enabling further configuration.

Reviewer: ```suggestion
[Kustomize](https://kustomize.io/) is a tool that allows you to modify Kubernetes YAML files without altering the original files. It can enhance the generated YAML from code-to-cloud deployments by applying additional customizations. The `kustomization.yaml` file in the root directory serves as a sample, combining the generated YAML files from different projects. Using Kustomize patches, you can add environment variables, such as specifying the location of the `Config.toml` file for a service, to enable additional configuration.
```

Reviewer: Is it correct to say `root directory serves as a **sample**`? Why do we say `sample` here?

Author: Remove sample word :  The `kustomization.yaml` file in the root directory demonstrates how to combine and manage the generated YAML files from multiple projects.

**kustomization.yaml**
```yaml
resources:
  - /path/to/target/kubernetes/hotel/hotel.ymal
  - /path/to/target/kubernetes/room/room.yaml
  - …

patchesStrategicMerge:
- secret-env-patch.yaml

```
See [Kustomize](https://github.com/ballerina-guides/gcp-microservices-demo?tab=readme-ov-file#kustomize) for more information.  

## CI/CD with Kubernetes

> This GitHub Action workflow automates the continuous integration and continuous deployment (CI/CD) process for a Ballerina project. It triggers every push to the repository, and builds the project.

Reviewer: ```suggestion
This GitHub Action workflow automates the continuous integration and continuous deployment (CI/CD) process for a Ballerina project. It triggers every push to the repository and builds the project.
```

    runs-on: ubuntu-latest

    steps:
      - name: Checkout
        uses: actions/checkout@v2

      # Step 2: Build the Ballerina Project
      - name: Ballerina Build
        uses: ballerina-platform/ballerina-action@master
        with:
          args: build --cloud=k8s

      # Step 3: Log in to Docker Registry
      - name: Log in to Docker Registry
        uses: docker/login-action@v2
        with:
          registry: ${{ secrets.DOCKER_REGISTRY_URL }}
          username: ${{ secrets.DOCKER_USERNAME }}
          password: ${{ secrets.DOCKER_PASSWORD }}

      # Step 4: Push the Docker Image
      - name: Push Docker Image
        run: |
          docker push ${{ secrets.DOCKER_REGISTRY_URL }}/your-docker-repo/your-app:${{ github.sha }}

      # Step 5: Set up Kubectl
      - name: Set up Kubectl
        uses: azure/setup-kubectl@v3
        with:
          version: 'v1.22.0'

      # Step 6: Deploy to Kubernetes
      - name: Deploy to Kubernetes
        env:
          KUBE_CONFIG_DATA: ${{ secrets.KUBE_CONFIG_DATA }}
        run: |
          echo "${KUBE_CONFIG_DATA}" | base64 --decode > $HOME/.kube/config
          kubectl apply -f <path to generated artifacts>
```

> - **Step 01:** The push event can initiate workflow to deployment process to the Kubernetes cluster.

Reviewer: ```suggestion
- **Step 01:** The push event can initiate a workflow to the deployment process to the Kubernetes cluster.
```

- **Step 02:** This step builds the Ballerina project, compiling the source code and preparing the necessary artifacts for deployment in Kubernetes.
- **Step 03:** This step logs into a Docker registry (such as Docker Hub or a private registry) to allow pushing the Docker image created in the build to the registry.
- **Step 04:** Pushes the Docker image to the specified Docker registry.
- **Step 05:** This step installs kubectl, the Kubernetes command-line tool, which is necessary for managing Kubernetes clusters and deploying applications.
- **Step 06:** Deploy the application to a Kubernetes cluster by applying the Kubernetes YAML configuration files generated by the Ballerina build in Step 2.
>**Note:** Replace <path to generated artifacts> with the actual path to the Kubernetes YAML files generated by the Ballerina build command (e.g., target/kubernetes).

> See [Code to Cloud deployment](https://ballerina.io/learn/code-to-cloud-deployment/) for more information on deployment options in Kubernetes and Docker.

Reviewer: We discussed to remove `Code to Cloud deployment` section.

Author: Yeah, we removed the sections and suggested using it as a reference here. May I change the page title as well?

Author: We removed this from the section. If we change the page name now, it could break links wherever it’s referenced. Therefore, shall we keep the name without change?
